fiftyone/core/odm/mixins.py

```diff
@@ -1552,7 +1552,11 @@ def _add_field_doc(field_docs, field_or_doc):
             field_docs[i] = new_field_doc
             return
 
-    field_docs.append(new_field_doc)
+    try:
+        field_docs.append(new_field_doc)
+    except ReferenceError:
+        # mongoengine seems to lose references to things... it's okay
```

Comment:
uh what? are we sure this isn't hiding some other problem?

Reply:
It's hiding something, but it's some weird mongoengine garbage collection issue that I don't understand.

Sit down for this one... without the `try-except` I added, the following code works if you paste the whole thing into ipython at once, but it raises an error if you wait one second between the `add_sample_field()` calls:

```py
import fiftyone as fo

dataset = fo.Dataset()

dataset.add_sample_field(
    "camera_info",
    fo.EmbeddedDocumentField,
    embedded_doc_type=fo.DynamicEmbeddedDocument,
)

# This fails if you wait one second between running this and the above code
dataset.add_sample_field("camera_info.camera_id", fo.StringField)
```

Reply:
I'm confident that everything works as expected if we just ignore this error. The last thing that happens at the end of `add_sample_field()` is a call to `dataset._reload()`, which recreates the relevant mongoengine classes and should paint over any weird issues due to mongoengine's `append()` implementation failing.

The unit tests I added verify that the dataset functions as intended.

Reply:
Thanks for the deep dive!